### ISSUE TITLE
Why sudo required for building image?

### ISSUE THREAD

**Issue by gww-parity** [project-oak/rust-verification-tools#139]
Why do you require sudo to build docker images?

I would love to test Klee but requireemnt to run `docker build` with sudo makes me suspicious.

**Comment by gww-parity** [comment]
Can we make some workaround to not require `sudo`?

**Comment by alastairreid** [comment]
On a Mac, you don't need sudo and, if you are used to managing packages with homebrew (which explicitly avoids sudo) then you will be disturbed by having to use sudo.

But, if you are on Linux then you need to use sudo with docker (and, of course, you would be using sudo with 'apt-get install') so there's no choice.

I don't know a good fix for this unfortunately.
I guess we would need to change

1) docker/mkimage
2) docker/run

which, I think, are the only occurences of sudo that execute on the host machine instead of in the docker container.

**Comment by gww-parity** [comment]
Do do I need sudo for building your image?

I build many images on daily basis without sudo.

Only yours require sudo.

**Comment by alastairreid** [comment]
Are you building on a Mac or a Linux machine?

I am not a docker expert but I was under the impression that you
needed sudo on a Linux machine.
If on a Mac, try removing sudo from the files I mentioned.
Let me know if that works.

--
Alastair

**Comment by gww-parity** [comment]
I am building on Linux machine.
I configured my system and Docker appropriately to not need sudo.
I have both machines, one with Archlinux and another with Ubuntu,
and I successfully configured on both systems that I don't need sudo to do `docker build`.
So far, I build all other Dockerfiles so far without sudo/root, from others and my own.
Only yours required sudo/root for some reason and when I removed `sudo` from your script it didn't work out.

**Comment by alastairreid** [comment]
Unfortunately, this is the first project I have ever used docker on and I
am doing so by a process of trial and error.
(This may have been obvious to you already :-))
If there is a (portable) way of using docker that avoids this problem, it
would be great to know.

**Comment by alsuren** [comment]
> If there is a (portable) way of using docker that avoids this problem, it
> would be great to know.

I opened #140 that *should* do what you want.


-- 

In separate but related news, I was looking into how `cross` manages to distribute pre-built images on dockerhub with only the root user, but avoids the problem of creating files with the wrong permissions on the host filesystem. After a bit of digging, I worked out that it does something like:

`docker run --userns host --rm --user 501:20 ...` where 501 is my userid on the mac, and 20 is the default "staff" group on the mac. Essentially it runs as in the host's user namespace, with the user/group from outside the the container. No need to set up the user within the docker image at all.

https://github.com/rust-embedded/cross/blob/master/src/docker.rs contains the business logic, and it might be reasonable to use it as inspiration if you wanted to make a `cargo-verify` wrapper that calls out to docker and publish it on crates.io.
